[k8s] Kubernetes Docs (#2324)
skypilot-org/skypilot: docs/source/reference/kubernetes/index.rst

.. _kubernetes-overview:

> Deploying on Kubernetes (Alpha)

Reviewer: Deploying sounds like "deploying skypilot (as a long-running service)". Wdyt about "Running on Kubernetes"?

Author: I'm ambivalent between "Deploying on Kubernetes" and "Running on Kubernetes". I've changed to "Running on Kubernetes".

Personally, I prefer "SkyPilot on Kubernetes", since it captures both - deployment (initial setup and considerations involved in SkyPilot on Kubernetes) and running (ongoing tasks, benefits, and features of using SkyPilot with Kubernetes) tasks. Ray also adopts a similar title ["Ray on Kubernetes"](https://docs.ray.io/en/latest/cluster/kubernetes/index.html).

.. _kubernetes-overview:

SkyPilot on Kubernetes (Alpha)
==============================

.. note::
    Kubernetes support for SkyPilot is an alpha preview under active development.
    There may be rough edges and features may change without notice.
    Please report any `bugs <https://github.com/skypilot-org/skypilot/issues>`_ and
    `reach out to us <http://slack.skypilot.co>`_ for feature requests.

SkyPilot tasks can be deployed on your private on-prem or cloud clusters running Kubernetes.
The Kubernetes cluster gets added to the list of "clouds" in SkyPilot and SkyPilot
tasks can be submitted to your Kubernetes cluster just like any other cloud provider.

> **Benefits of bringing your Kubernetes cluster to SkyPilot:**

Reviewer: ```suggestion
**Benefits of using SkyPilot to run jobs on your Kubernetes cluster:**
```
This reads better to me. Wdyt?

* Get SkyPilot features (setup management, job execution, queuing, logging, SSH access) on your Kubernetes resources
* Replace complex Kubernetes manifests with simple SkyPilot tasks

> * Maximize resource utilization by running cloud jobs on your Kubernetes cluster.

Reviewer: Not sure I get this part. Should we leave it out?

> * Seamlessly "burst" jobs to the cloud if the Kubernetes cluster is congested.

Reviewer: ```suggestion
* Seamlessly "burst" jobs to the cloud if the Kubernetes cluster is congested
```

> * Retain observability and control over your cluster with your existing Kubernetes tools.

Reviewer: ```suggestion
* Retain observability and control over your cluster with your existing Kubernetes tools
```

**Supported deployment models:**

> * On-prem clusters (Kubeadm, K3s, Rancher)
> * Hosted Kubernetes services (AWS EKS, GKE)
> * Local development clusters (KinD, minikube)

Reviewer: ```suggestion
* Hosted Kubernetes services (AWS EKS, GCP GKE)
* On-prem clusters (Kubeadm, K3s, Rancher)
* Local development clusters (KinD, minikube)
```
reordered to put most common/targeted at top

Kubernetes Cluster Requirements
-------------------------------

To connect and use a Kubernetes cluster, SkyPilot needs:

* An existing Kubernetes cluster running Kubernetes v1.20 or later.

> * A `Kubeconfig <kubeconfig>`_ file containing access credentials and namespace to be used.

Reviewer: link doesn't work?

In a typical workflow:

1. A cluster administrator sets up a Kubernetes cluster. Detailed admin guides for
   different deployment environments (Amazon EKS, Google GKE, On-Prem and local debugging) are included in the :ref:`Kubernetes cluster setup guide <kubernetes-setup>`.

2. Users who want to run SkyPilot tasks on this cluster are issued Kubeconfig
   files containing their credentials (`kube-context <https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#define-clusters-users-and-contexts>`_).
   SkyPilot reads this Kubeconfig file to communicate with the cluster.

Submitting SkyPilot tasks to Kubernetes Clusters
------------------------------------------------
.. _kubernetes-instructions:

Once your cluster administrator has :ref:`setup a Kubernetes cluster <kubernetes-setup>` and provided you with a kubeconfig file:

> 0. Make sure `kubectl <https://kubernetes.io/docs/tasks/tools/>`_, ``socat`` and ``lsof`` are installed.

Reviewer: Where? on local machine? My mac doesn't have `socat` but it ran k8s before.

Author: Ah thanks - clarified for local machine now. `socat` and `lsof` are dependencies introduced with #2500.

Reviewer: Not sure what the best way but wearing a user's hat, having `socat` as a requirement is barrier to entry? Mac doesn't come with it. Can we skip it from docs?

Author: It's required, so cannot be removed from docs. I think it's a fairly low barrier to entry, and about as hard as installing kubectl (i.e., one line `brew install kubectl socat`). I've updated the docs to include commands for installing these dependencies. 

Our code also includes pre-flight checks for these deps, and if not present throws a nice error suggesting how to install.

1. Place your kubeconfig file at ``~/.kube/config``.

   .. code-block:: console

     $ mkdir -p ~/.kube
     $ cp /path/to/kubeconfig ~/.kube/config

>    You can verify your credentials are setup correctly by running :code:`kubectl get pods`

Reviewer: ```suggestion
   You can verify your credentials are setup correctly by running :code:`kubectl get pods`.
```

         Kubernetes   2CPU--2GB         2       2         -              kubernetes    0.00          ✔
         AWS          m6i.large         2       8         -              us-east-1     0.10
         Azure        Standard_D2s_v5   2       8         -              eastus        0.10
         GCP          n2-standard-2     2       8         -              us-central1   0.10
         IBM          bx2-8x32          8       32        -              us-east       0.38
         Lambda       gpu_1x_a10        30      200       A10:1          us-east-1     0.60
        ---------------------------------------------------------------------------------------------------.

.. note::
  SkyPilot will use the cluster and namespace set in the ``current-context`` in the
  kubeconfig file. To manage your ``current-context``:

  .. code-block:: console

    $ # See current context
    $ kubectl config current-context

    $ # Switch current-context
    $ kubectl config use-context mycontext

    $ # Set a specific namespace to be used in the current-context
    $ kubectl config set-context --current --namespace=mynamespace

FAQs
----

* **Are autoscaling Kubernetes clusters supported?**

>   Yes - however they currently require adjusting the resource provisioning timeout (:code:`Kubernetes.TIMEOUT` in `clouds/kubernetes.py`) to a large value to give enough time for the cluster to autoscale. We are working on a better interface to adjust this timeout - stay tuned!

Reviewer: "Yes" -> "Partially. To use it, they require ..."?

Reason: if the default doesn't work for common autoscaling setup then it sounds more like partially?

Not feeling strongly about this one.

Author: Reworded a bit, wdyt?

* **What container image is used for tasks? Can I specify my own image?**

  We use and maintain a SkyPilot container image that has conda and a few other basic tools installed. You can specify a custom image to use in `clouds/kubernetes.py`, but it must have rsync, conda and OpenSSH server installed. We are working on a interface to allow specifying custom images through the :code:`image_id` field in the task YAML - stay tuned!

* **Can SkyPilot provision a Kubernetes cluster for me? Will SkyPilot add more nodes to my Kubernetes clusters?**

>   The goal of SkyPilot on Kubernetes is to run SkyPilot tasks on resources in an existing Kubernetes cluster. It does not provision any new Kubernetes clusters or add new nodes to an existing Kubernetes cluster. The Kubernetes control plane remains untouched.

Reviewer: ```suggestion
  The goal of Kubernetes support is to run SkyPilot tasks in an existing Kubernetes cluster. It does not provision any new Kubernetes clusters or add new nodes to an existing Kubernetes cluster. The Kubernetes control plane remains untouched.
```
"The Kubernetes control plane remains untouched." this part reads a bit surprising under this Q.

* **I have multiple users in my organization who share the same Kubernetes cluster. How do I provide isolation for their SkyPilot workloads?**

>   You can create an individual Kubernetes namespace for each user and set it in the kubeconfig distributed to the user. SkyPilot will use this namespace for running all tasks.

Reviewer: Is per user common, or per team/project? The former reads surprising. Perhaps add the latter as an option too?

Author: Thanks - updated the language to avoid mentioning specific isolation units (user/team/project/org)

Features and Roadmap
--------------------

> SkyPilot on Kubernetes is under active development. Some features are in progress and will be released soon:

Reviewer: ```suggestion
Kubernetes support is under active development. Some features are in progress and will be released soon:
```

* CPU Tasks - ✅ Available
* Auto-down - ✅ Available
* Storage mounting - ✅ Available on x86_64 clusters

> * GPU Tasks - ✅ Available

Reviewer: move to top?

Author: I'd prefer to prioritize other information first - the benefits of using SkyPilot on k8s and how to get started. wdyt?

Reviewer: I meant that L146 GPU Tasks should be moved to the top of the bullet list.

.. _kubernetes-overview:

Deploying on Kubernetes (Alpha)
======================================

.. note::
    Kubernetes support is in alpha preview and under active development.
    There may be rough edges and features may change without notice.
    Please report any `bugs <https://github.com/skypilot-org/skypilot/issues>`_ and
    `reach out to us <http://slack.skypilot.co>`_ for feature requests.

> SkyPilot tasks can be deployed on your private on-prem or cloud Kubernetes clusters.

Reviewer: ```suggestion
SkyPilot tasks can be run on your private on-prem or cloud Kubernetes clusters.
```
or executed?

Deploy usually refers to long-running services.

The Kubernetes cluster gets added to the list of "clouds" in SkyPilot and SkyPilot
tasks can be submitted to your Kubernetes cluster just like any other cloud provider.

**Benefits of using SkyPilot to run jobs on your Kubernetes cluster:**

> * Get SkyPilot features (setup management, job execution, queuing, logging, SSH access) on your Kubernetes resources

Reviewer: ```suggestion
* Get ergonomic SkyPilot features (setup management, job execution, queuing, logging, SSH access) on your Kubernetes resources
```

Author: Not sure about adding ergonomic - sounds a little fluffy. The convenience is covered in the examples in brackets and the next bullet point.

* Replace complex Kubernetes manifests with simple SkyPilot tasks
* Seamlessly "burst" jobs to the cloud if your Kubernetes cluster is congested
* Retain observability and control over your cluster with your existing Kubernetes tools

> **Supported deployment models:**

Reviewer: ```suggestion
**Supported Kubernetes deployments:**
```

   .. code-block:: console

>         $ sky launch --cpus 2+

Reviewer: nit: is task.yaml optional?

Author: Yes, it's optional - `sky launch --cpus 2+` also works. However, I added task.yaml to avoid confusing users ("Where do I specify my task yaml? Is Kubernetes any different from regular SkyPilot in this regard?").

* **Are autoscaling Kubernetes clusters supported?**

>   To deploy on an autoscaling cluster, you may need to adjust the resource provisioning timeout (:code:`Kubernetes.TIMEOUT` in `clouds/kubernetes.py`) to a large value to give enough time for the cluster to autoscale. We are working on a better interface to adjust this timeout - stay tuned!

Reviewer: ```suggestion
  To run on an autoscaling cluster, you may need to adjust the resource provisioning timeout (:code:`Kubernetes.TIMEOUT` in `clouds/kubernetes.py`) to a large value to give enough time for the cluster to autoscale. We are working on a better interface to adjust this timeout - stay tuned!
```
